packages/astro/src/content/vite-plugin-content-assets.ts

```diff
@@ -48,52 +42,59 @@ export function astroContentAssetPropagationPlugin({
 				return this.resolve(base, importer, { skipSelf: true, ...opts });
 			}
 		},
-		configureServer(server) {
-			if (mode === 'dev') {
-				devModuleLoader = createViteLoader(server);
-			}
-		},
 		async transform(_, id, options) {
 			if (hasContentFlag(id, PROPAGATED_ASSET_FLAG)) {
 				const basePath = id.split('?')[0];
 				let stringifiedLinks: string, stringifiedStyles: string, stringifiedScripts: string;
 
+
+				let code = ''
+
 				// We can access the server in dev,
 				// so resolve collected styles and scripts here.
-				if (options?.ssr && devModuleLoader) {
-					if (!devModuleLoader.getModuleById(basePath)?.ssrModule) {
-						await devModuleLoader.import(basePath);
-					}
-					const { styles, urls } = await getStylesForURL(pathToFileURL(basePath), devModuleLoader);
+				if (options?.ssr && mode === 'dev') {
+					code = `import devLoader from 'astro:dev-module-loader';
+import { createGetPropagatedAssets } from 'astro/virtual-modules/dev-helpers.js';
+
+const getPropagatedAssets = createGetPropagatedAssets(devLoader, ${JSON.stringify(basePath)}, ${JSON.stringify(settings.config.root)});
 
-					const hoistedScripts = await getScriptsForURL(
-						pathToFileURL(basePath),
-						settings.config.root,
-						devModuleLoader
-					);
+async function getMod() {
+	return import(${JSON.stringify(basePath)});
+}
 
-					stringifiedLinks = JSON.stringify([...urls]);
-					stringifiedStyles = JSON.stringify(styles.map((s) => s.content));
-					stringifiedScripts = JSON.stringify([...hoistedScripts]);
+const defaultMod = { __astroPropagation: true, getMod, getPropagatedAssets };
+export default defaultMod;
+`;
 				} else {
 					// Otherwise, use placeholders to inject styles and scripts
 					// during the production bundle step.
 					// @see the `astro:content-build-plugin` below.
 					stringifiedLinks = JSON.stringify(LINKS_PLACEHOLDER);
 					stringifiedStyles = JSON.stringify(STYLES_PLACEHOLDER);
 					stringifiedScripts = JSON.stringify(SCRIPTS_PLACEHOLDER);
+
+					code = `
+async function getMod() {
+	return import(${JSON.stringify(basePath)});
+}
+
+const collectedLinks = ${stringifiedLinks};
+const collectedStyles = ${stringifiedStyles};
+const collectedScripts = ${stringifiedScripts};
+
+function getPropagatedAssets() {
```

Comment:
Dang. Clever solution! Not _too_ clever of course, just the right amount of clever. 👌